neo/Consensus/RecoveryMessage.cs

```diff
@@ -0,0 +1,103 @@
+﻿using System;
+using System.IO;
+using System.Linq;
+using Neo.IO;
+using Neo.Network.P2P.Payloads;
+
+namespace Neo.Consensus
+{
+    internal class RecoveryMessage : ConsensusMessage
```

Comment:
The `RecoveryMessage` holds the necessary data for recovering the `ChangeView`, `PrepareRequest`, `PrepareResponse`, and `Commit` messages.

If a node has not yet received a `PrepareRequest`, it should not send data other than `ChangeView` when sending a `RecoveryMessage` because the data is unverified.

Reply:
It doesn't hurt to receive it though; it is just like when you receive the `PrepareResponse` messages before the `PrepareRequest` when network messages come out of order.

Reply:
The `PrepareRequest` that goes along with it can come in a subsequent message and it will recover faster the way it is.

Reply:
If you didn't receive `PrepareRequest`, and the `PreparationHash` in `PrepareResponse` you received are different with each other, how do you set the `RecoveryMessage.PreparationHash`?

Reply:
Right now it just takes the first one that isn't null from the Preparations and throws it in the `RecoveryMessage.PreparationHash`. They are most likely all equal to it anyway; any that aren't equal to it will just fail to validate when it handles the recovery message. It only uses the `RecoveryMessage.PreparationHash` when it doesn't have the `PrepareRequest` to serialize into the` RecoveryMessage`.

Reply:
Accepted your improvement to take the most common hash instead of the first one.

Reply:
It probably should null the other ones that don’t match out to serialize less data too, and not try to verify the ones that are just going to fail verification.